sources/udevd-164/udevd-164/udev/udevadm.8
'\" t
.\"     Title: udevadm
.\"    Author: [see the "Author" section]
.\" Generator: DocBook XSL Stylesheets v1.75.2 <http://docbook.sf.net/>
.\"      Date: 10/23/2010
.\"    Manual: udevadm
.\"    Source: udev
.\"  Language: English
.\"
.TH "UDEVADM" "8" "10/23/2010" "udev" "udevadm"
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
udevadm \- udev management tool
.SH "SYNOPSIS"
.HP \w'\fBudevadm\fR\ 'u
\fBudevadm\fR [\fB\-\-debug\fR] [\fB\-\-version\fR] [\fB\-\-help\fR]
.HP \w'\fBudevadm\ info\ \fR\fB\fIoptions\fR\fR\ 'u
\fBudevadm info \fR\fB\fIoptions\fR\fR
.HP \w'\fBudevadm\ trigger\ \fR\fB[options]\fR\ 'u
\fBudevadm trigger \fR\fB[options]\fR
.HP \w'\fBudevadm\ settle\ \fR\fB[options]\fR\ 'u
\fBudevadm settle \fR\fB[options]\fR
.HP \w'\fBudevadm\ control\ \fR\fB\fIcommand\fR\fR\ 'u
\fBudevadm control \fR\fB\fIcommand\fR\fR
.HP \w'\fBudevadm\ monitor\ \fR\fB[options]\fR\ 'u
\fBudevadm monitor \fR\fB[options]\fR
.HP \w'\fBudevadm\ test\ \fR\fB[options]\fR\fB\ \fR\fB\fIdevpath\fR\fR\ 'u
\fBudevadm test \fR\fB[options]\fR\fB \fR\fB\fIdevpath\fR\fR
.SH "DESCRIPTION"
.PP
udevadm expects a command and command specific options\&. It controls the runtime behavior of udev, requests kernel events, manages the event queue, and provides simple debugging mechanisms\&.
.SH "OPTIONS"
.PP
\fB\-\-debug\fR
.RS 4
Print debug messages to stderr\&.
.RE
.PP
\fB\-\-version\fR
.RS 4
Print version number\&.
.RE
.PP
\fB\-\-help\fR
.RS 4
Print help text\&.
.RE
.SS "udevadm info \fIoptions\fR"
.PP
Queries the udev database for device information stored in the udev database\&. It can also query the properties of a device from its sysfs representation to help creating udev rules that match this device\&.
.PP
\fB\-\-query=\fR\fB\fItype\fR\fR
.RS 4
Query the database for specified type of device data\&. It needs the
\fB\-\-path\fR
or
\fB\-\-name\fR
to identify the specified device\&. Valid queries are:
\fBname\fR,
\fBsymlink\fR,
\fBpath\fR,
\fBproperty\fR,
\fBall\fR\&.
.RE
.PP
\fB\-\-path=\fR\fB\fIdevpath\fR\fR
.RS 4
The devpath of the device to query\&.
.RE
.PP
\fB\-\-name=\fR\fB\fIfile\fR\fR
.RS 4
The name of the device node or a symlink to query
.RE
.PP
\fB\-\-root\fR
.RS 4
The udev root directory:
/dev\&. If used in conjunction with a
\fBname\fR
or
\fBsymlink\fR
query, the query returns the absolute path including the root directory\&.
.RE
.PP
\fB\-\-attribute\-walk\fR
.RS 4
Print all sysfs properties of the specified device that can be used in udev rules to match the specified device\&. It prints all devices along the chain, up to the root of sysfs that can be used in udev rules\&.
.RE
.PP
\fB\-\-device\-id\-of\-file=\fR\fB\fIfile\fR\fR
.RS 4
Print major/minor numbers of the underlying device, where the file lives on\&.
.RE
.PP
\fB\-\-export\-db\fR
.RS 4
Export the content of the udev database\&.
.RE
.PP
\fB\-\-version\fR
.RS 4
Print version\&.
.RE
.PP
\fB\-\-help\fR
.RS 4
Print help text\&.
.RE
.SS "udevadm trigger [options]"
.PP
Request device events from the kernel\&. Primarily used to replay events at system coldplug time\&.
.PP
\fB\-\-verbose\fR
.RS 4
Print the list of devices which will be triggered\&.
.RE
.PP
\fB\-\-dry\-run\fR
.RS 4
Do not actually trigger the event\&.
.RE
.PP
\fB\-\-type=\fR\fB\fItype\fR\fR
.RS 4
Trigger a specific type of devices\&. Valid types are:
\fBdevices\fR,
\fBsubsystems\fR,
\fBfailed\fR\&. The default value is
\fBdevices\fR\&.
.RE
.PP
\fB\-\-action=\fR\fB\fIaction\fR\fR
.RS 4
Type of event to be triggered\&. The default value is
\fBchange\fR\&.
.RE
.PP
\fB\-\-subsystem\-match=\fR\fB\fIsubsystem\fR\fR
.RS 4
Trigger events for devices which belong to a matching subsystem\&. This option can be specified multiple times and supports shell style pattern matching\&.
.RE
.PP
\fB\-\-subsystem\-nomatch=\fR\fB\fIsubsystem\fR\fR
.RS 4
Do not trigger events for devices which belong to a matching subsystem\&. This option can be specified multiple times and supports shell style pattern matching\&.
.RE
.PP
\fB\-\-attr\-match=\fR\fB\fIattribute\fR\fR\fB=\fR\fB\fIvalue\fR\fR
.RS 4
Trigger events for devices with a matching sysfs attribute\&. If a value is specified along with the attribute name, the content of the attribute is matched against the given value using shell style pattern matching\&. If no value is specified, the existence of the sysfs attribute is checked\&. This option can be specified multiple times\&.
.RE
.PP
\fB\-\-attr\-nomatch=\fR\fB\fIattribute\fR\fR\fB=\fR\fB\fIvalue\fR\fR
.RS 4
Do not trigger events for devices with a matching sysfs attribute\&. If a value is specified along with the attribute name, the content of the attribute is matched against the given value using shell style pattern matching\&. If no value is specified, the existence of the sysfs attribute is checked\&. This option can be specified multiple times\&.
.RE
.PP
\fB\-\-property\-match=\fR\fB\fIproperty\fR\fR\fB=\fR\fB\fIvalue\fR\fR
.RS 4
Trigger events for devices with a matching property value\&. This option can be specified multiple times and supports shell style pattern matching\&.
.RE
.PP
\fB\-\-tag\-match=\fR\fB\fIproperty\fR\fR
.RS 4
Trigger events for devices with a matching tag\&. This option can be specified multiple times\&.
.RE
.PP
\fB\-\-sysname\-match=\fR\fB\fIname\fR\fR
.RS 4
Trigger events for devices with a matching sys device name\&. This option can be specified multiple times and supports shell style pattern matching\&.
.RE
.SS "udevadm settle [options]"
.PP
Watches the udev event queue, and exits if all current events are handled\&.
.PP
\fB\-\-timeout=\fR\fB\fIseconds\fR\fR
.RS 4
Maximum number of seconds to wait for the event queue to become empty\&. The default value is 180 seconds\&. A value of 0 will check if the queue is empty and always return immediately\&.
.RE
.PP
\fB\-\-seq\-start=\fR\fB\fIseqnum\fR\fR
.RS 4
Wait only for events after the given sequence number\&.
.RE
.PP
\fB\-\-seq\-end=\fR\fB\fIseqnum\fR\fR
.RS 4
Wait only for events before the given sequence number\&.
.RE
.PP
\fB\-\-exit\-if\-exists=\fR\fB\fIfile\fR\fR
.RS 4
Stop waiting if file exists\&.
.RE
.PP
\fB\-\-quiet\fR
.RS 4
Do not print any output, like the remaining queue entries when reaching the timeout\&.
.RE
.PP
\fB\-\-help\fR
.RS 4
Print help text\&.
.RE
.SS "udevadm control \fIcommand\fR"
.PP
Modify the internal state of the running udev daemon\&.
.PP
\fB\-\-log\-priority=\fR\fB\fIvalue\fR\fR
.RS 4
Set the internal log level of udevd\&. Valid values are the numerical syslog priorities or their textual representations:
\fBerr\fR,
\fBinfo\fR
and
\fBdebug\fR\&.
.RE
.PP
\fB\-\-stop\-exec\-queue\fR
.RS 4
Signal udevd to stop executing new events\&. Incoming events will be queued\&.
.RE
.PP
\fB\-\-start\-exec\-queue\fR
.RS 4
Signal udevd to enable the execution of events\&.
.RE
.PP
\fB\-\-reload\-rules\fR
.RS 4
Signal udevd to reload the rules files\&. The udev daemon detects changes automatically, this option is usually not needed\&. Reloading rules does not apply any changes to already existing devices\&.
.RE
.PP
\fB\-\-property=\fR\fB\fIKEY\fR\fR\fB=\fR\fB\fIvalue\fR\fR
.RS 4
Set a global property for all events\&.
.RE
.PP
\fB\-\-children\-max=\fR\fIvalue\fR
.RS 4
Set the maximum number of events, udevd will handle at the same time\&.
.RE
.PP
\fB\-\-help\fR
.RS 4
Print help text\&.
.RE
.SS "udevadm monitor [options]"
.PP
Listens to the kernel uevents and events sent out by a udev rule and prints the devpath of the event to the console\&. It can be used to analyze the event timing, by comparing the timestamps of the kernel uevent and the udev event\&.
.PP
\fB\-\-kernel\fR
.RS 4
Print the kernel uevents\&.
.RE
.PP
\fB\-\-udev\fR
.RS 4
Print the udev event after the rule processing\&.
.RE
.PP
\fB\-\-property\fR
.RS 4
Also print the properties of the event\&.
.RE
.PP
\fB\-\-subsystem\-match=\fR\fB\fIstring[/string]\fR\fR
.RS 4
Filter events by subsystem[/devtype]\&. Only udev events with a matching subsystem value will pass\&.
.RE
.PP
\fB\-\-tag\-match=\fR\fB\fIstring\fR\fR
.RS 4
Filter events by property\&. Only udev events with a given tag attached will pass\&.
.RE
.PP
\fB\-\-help\fR
.RS 4
Print help text\&.
.RE
.SS "udevadm test [options] \fIdevpath\fR"
.PP
Simulate a udev event run for the given device, and print debug output\&.
.PP
\fB\-\-action=\fR\fB\fIstring\fR\fR
.RS 4
The action string\&.
.RE
.PP
\fB\-\-subsystem=\fR\fB\fIstring\fR\fR
.RS 4
The subsystem string\&.
.RE
.PP
\fB\-\-help\fR
.RS 4
Print help text\&.
.RE
.SH "AUTHOR"
.PP
Written by Kay Sievers
kay\&.sievers@vrfy\&.org\&.
.SH "SEE ALSO"
.PP
\fBudev\fR(7)
\fBudevd\fR(8)


sources/udevd-164/udevd-164/udev/udevadm.c
/*
 * Copyright (C) 2007-2009 Kay Sievers <kay.sievers@vrfy.org>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#include <errno.h>
#include <getopt.h>

#include "udev.h"

static int debug;

static void log_fn(struct udev *udev, int priority,
		   const char *file, int line, const char *fn,
		   const char *format, va_list args)
{
	if (debug) {
		fprintf(stderr, "%s: ", fn);
		vfprintf(stderr, format, args);
	} else {
		va_list args2;

		va_copy(args2, args);
		vfprintf(stderr, format, args2);
		va_end(args2);
		vsyslog(priority, format, args);
	}
}

struct command {
	const char *name;
	int (*cmd)(struct udev *udev, int argc, char *argv[]);
	const char *help;
	int debug;
};

static const struct command cmds[];

static int version(struct udev *udev, int argc, char *argv[])
{
	printf("%s\n", VERSION);
	return 0;
}

static int help(struct udev *udev, int argc, char *argv[])
{
	const struct command *cmd;

	printf("Usage: udevadm [--help] [--version] [--debug] COMMAND [COMMAND OPTIONS]\n");
	for (cmd = cmds; cmd->name != NULL; cmd++)
		if (cmd->help != NULL)
			printf("  %-12s %s\n", cmd->name, cmd->help);
	printf("\n");
	return 0;
}

static const struct command cmds[] = {
	{
		.name = "info",
		.cmd = udevadm_info,
		.help = "query sysfs or the udev database",
	},
	{
		.name = "trigger",
		.cmd = udevadm_trigger,
		.help = "request events from the kernel",
	},
	{
		.name = "settle",
		.cmd = udevadm_settle,
		.help = "wait for the event queue to finish",
	},
	{
		.name = "control",
		.cmd = udevadm_control,
		.help = "control the udev daemon",
	},
	{
		.name = "monitor",
		.cmd = udevadm_monitor,
		.help = "listen to kernel and udev events",
	},
	{
		.name = "test",
		.cmd = udevadm_test,
		.help = "simulation run",
		.debug = 1,
	},
	{
		.name = "version",
		.cmd = version,
	},
	{
		.name = "help",
		.cmd = help,
	},
	{
		.name = NULL,
		.cmd = NULL,
		.help = NULL,
		.debug = 0,
	}
};

static int run_command(struct udev *udev, const struct command *cmd, int argc, char *argv[])
{
	if (cmd->debug) {
		debug = 1;
		if (udev_get_log_priority(udev) < LOG_INFO)
			udev_set_log_priority(udev, LOG_INFO);
	}
	info(udev, "calling: %s\n", cmd->name);
	return cmd->cmd(udev, argc, argv);
}

int main(int argc, char *argv[])
{
	struct udev *udev;
	static const struct option options[] = {
		{ "debug", no_argument, NULL, 'd' },
		{ "help", no_argument, NULL, 'h' },
		{ "version", no_argument, NULL, 'V' },
		{ NULL, 0, NULL, 0}
	};
	const char *command;
	int i;
	int rc = 1;

	udev = udev_new();
	if (udev == NULL)
		goto out;

	udev_log_init("udevadm");
	udev_set_log_fn(udev, log_fn);
	udev_selinux_init(udev);

	for (;;) {
		int option;

		option = getopt_long(argc, argv, "+dhV", options, NULL);
		if (option == -1)
			break;

		switch (option) {
		case 'd':
			debug = 1;
			if (udev_get_log_priority(udev) < LOG_INFO)
				udev_set_log_priority(udev, LOG_INFO);
			break;
		case 'h':
			rc = help(udev, argc, argv);
			goto out;
		case 'V':
			rc = version(udev, argc, argv);
			goto out;
		default:
			goto out;
		}
	}
	command = argv[optind];

	if (command != NULL)
		for (i = 0; cmds[i].cmd != NULL; i++) {
			if (strcmp(cmds[i].name, command) == 0) {
				argc -= optind;
				argv += optind;
				optind = 0;
				rc = run_command(udev, &cmds[i], argc, argv);
				goto out;
			}
		}

	fprintf(stderr, "missing or unknown command\n\n");
	help(udev, argc, argv);
	rc = 2;
out:
	udev_selinux_exit(udev);
	udev_unref(udev);
	udev_log_close();
	return rc;
}
